homeassistant/components/sensor/mvglive.py

```diff
@@ -0,0 +1,160 @@
+        self._station = station
+        self._destination = destination
+        self._line = line
+        self.data = MVGLiveData(station, destination, line,
+                                offset, ubahn, tram, bus, sbahn)
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        # e.g.
+        # 'Hauptbahnhof (S1)'
+        # 'Hauptbahnhof-Marienplatz'
+        # 'Hauptbahnhof-Marienplatz (S1)'
+        namestr = self._station
+        if self._destination:
+            namestr = namestr + '-' + self._destination
+        if self._line:
+            namestr = namestr + ' (' + self._line + ')'
+        return namestr
+
+    @property
+    def icon(self):
+        """Return the icon for the frontend."""
+        return ICON
+
+    @property
+    def state(self):
+        """Return the departure time of the next train."""
+        return self._state
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        if self.data.departures:
+            return self.data.departures[0]
+
+    def update(self):
+        """Get the latest data and update the state."""
+        self.data.update()
+        if not self.data.departures:
+            self._state = ('-')
+        else:
+            self._state = self.data.departures[0].get('time', '-')
+
+
+class MVGLiveData(object):
+    """Pull data from the mvg-live.de web page."""
+
+    def __init__(self, station, destination, line,
+                 offset, ubahn, tram, bus, sbahn):
+        """Initialize the sensor."""
+        import MVGLive
+        self._station = station
+        self._destination = destination
+        self._line = line
+        self._offset = offset
+        self._ubahn = ubahn
+        self._tram = tram
+        self._bus = bus
+        self._sbahn = sbahn
+        self.mvg = MVGLive.MVGLive()
+        self.departures = [{}]
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Update the connection data."""
+        try:
+            self.departures = self.mvg.getlivedata(station=self._station,
+                                                   ubahn=self._ubahn,
+                                                   tram=self._tram,
+                                                   bus=self._bus,
+                                                   sbahn=self._sbahn)
+        except ValueError:
+            self.departures = [{}]
+            _LOGGER.warning("Returned data not understood.")
+            return
+        self.departures = [con for con in self.departures
+                           if con['time'] >= self._offset and
+                           con['destination'].startswith(self._destination)]
+        for con in self.departures:
+            # Details info is not useful.
+            # Having a more consistent interface simplifies
+            # usage of Template sensors later on
+            con.pop('productsymbol')
```

Comment:
It's usually not a good idea to mutate data that you fetch from a library. They might store a reference in a cache that is now no longer what they expect it to be. Create new dictionaries instead.